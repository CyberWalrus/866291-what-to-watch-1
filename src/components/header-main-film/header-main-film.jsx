import React, {Fragment} from "react";
import PropTypes from "prop-types";
import Header from "../header/header.jsx";
import {connect} from "react-redux";
import {getFilm} from "../../store/data/selectors.js";

const HeaderMainFilm = ({film, onVideoScreenOpen}) => {
  if (film) {
    return (
      <Fragment>
        <section className="movie-card">
          <div className="movie-card__bg">
            <img src={film.srcBgImage} alt={film.title} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <Header />

          <div className="movie-card__wrap">
            <div className="movie-card__info">
              <div className="movie-card__poster">
                <img
                  src={film.srcPosterImage}
                  alt={`${film.title} poster`}
                  width="218"
                  height="327"
                />
              </div>

              <div className="movie-card__desc">
                <h2 className="movie-card__title">{film.title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{film.genre}</span>
                  <span className="movie-card__year">{film.released}</span>
                </p>

                <div className="movie-card__buttons">
                  <button
                    className="btn btn--play movie-card__button"
                    type="button"
                    onClick={onVideoScreenOpen}
                  >
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s" />
                    </svg>
                    <span>Play</span>
                  </button>
                  <button
                    className="btn btn--list movie-card__button"
                    type="button"
                  >
                    <svg viewBox="0 0 19 20" width="19" height="20">
                      <use xlinkHref="#add" />
                    </svg>
                    <span>My list</span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </section>
      </Fragment>
    );
  }
  return <Header />;
};
HeaderMainFilm.propTypes = {
  onVideoScreenOpen: PropTypes.func.isRequired,
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    bgColor: PropTypes.string.isRequired,
    srcPreviewImage: PropTypes.string.isRequired,
    srcPosterImage: PropTypes.string.isRequired,
    srcBgImage: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    released: PropTypes.number.isRequired,
    preview: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    scoresCount: PropTypes.number.isRequired,
    ratingLevel: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    starrings: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    runTime: PropTypes.number.isRequired
  })
};
const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    film: getFilm(state, 1)
  });

export {HeaderMainFilm};

export default connect(mapStateToProps)(HeaderMainFilm);
