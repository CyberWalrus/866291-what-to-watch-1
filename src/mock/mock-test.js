const FILMS = [
  {
    bgColor: `#977461`,
    description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
    director: `Martin Scorsese`,
    genre: `Thriller`,
    id: 1,
    isFavorite: false,
    pageUrl: `Shutter Island`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: `8,1`,
    ratingLevel: `Very good`,
    released: 2010,
    runTime: 138,
    scoresCount: 1002557,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
    starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
    title: `Shutter Island`
  },
  {
    bgColor: `#F1E9CE`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    pageUrl: `Macbeth`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: `6,6`,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    starrings: [(`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`)],
    title: `Macbeth`
  }
];
const FILM = {
  bgColor: `#977461`,
  description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
  director: `Martin Scorsese`,
  genre: `Thriller`,
  id: 1,
  isFavorite: false,
  pageUrl: `Shutter Island`,
  srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: `8,1`,
  ratingLevel: `Very good`,
  released: 2010,
  runTime: 138,
  scoresCount: 1002557,
  srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
  srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
  srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
  starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
  title: `Shutter Island`
};
const REVIEWS = [
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  },
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  }
];

const GENRES = [
  `All genres`,
  `Thriller`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Action`,
  `Fantasy`
];
export {FILMS, FILM, GENRES, REVIEWS};
