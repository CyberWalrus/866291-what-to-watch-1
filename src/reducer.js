import films from "./mock/films.js";
const intialState = {
  genreFilter: `all`,
  films
};

const ActionCreator = {
  changeGenre: (value) => {
    if (
      [
        `comedy`,
        `crime`,
        `documentary`,
        `drama`,
        `horror`,
        `family`,
        `romance`,
        `sci-fi`,
        `thriller`
      ].find((item) => item === value)
    ) {
      const filmsNew = films.filter((item) => item.genre === value);
      return {
        type: `CHANGE_GENRE`,
        genreNew: value,
        filmsNew
      };
    }
    return {
      type: `RESET`
    };
  }
};

const reducer = (state = intialState, action) => {
  switch (action.type) {
    case `CHANGE_GENRE`:
      return Object.assign({}, state, {
        genreFilter: action.genreNew,
        films: action.filmsNew
      });
    case `RESET`:
      return Object.assign({}, intialState);
  }
  return state;
};

export {ActionCreator, reducer};
