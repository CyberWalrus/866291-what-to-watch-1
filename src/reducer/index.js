import {combineReducers} from "redux";
import {reducer as filter} from "./filter/filter.js";
import {reducer as data} from "./data/data.js";
import NameSpace from "./name-spaces.js";

export default combineReducers({
  [NameSpace.FILTER]: filter,
  [NameSpace.DATA]: data
});
