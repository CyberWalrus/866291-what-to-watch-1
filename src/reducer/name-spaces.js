const NameSpace = {
  DATA: `DATA`,
  FILTER: `FILTER`
};

export default NameSpace;
