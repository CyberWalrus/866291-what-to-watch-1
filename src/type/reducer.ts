import {ThunkDispatch as ReduxThunkDispatch, ThunkAction as ReduxThunkAction} from "redux-thunk";
import {AxiosInstance} from "axios";
import {State as StateData, Action as ActionData} from "../store/data/data";
import {State as StateFilter, Action as ActionFilter} from "../store/filter/filter";
import {State as StateUser, Action as ActionUser} from "../store/user/user";
import NameSpace from "../store/name-spaces";

export interface StateApp {
  [NameSpace.DATA]: StateData;
  [NameSpace.FILTER]: StateFilter;
  [NameSpace.USER]: StateUser;
}

export type ActionApp = ActionData | ActionFilter | ActionUser;
export type ThunkDispatch = ReduxThunkDispatch<StateApp, AxiosInstance, ActionApp>;
export type ThunkAction = ReduxThunkAction<
Promise<void>,
StateApp,
AxiosInstance,
ActionApp
>;
