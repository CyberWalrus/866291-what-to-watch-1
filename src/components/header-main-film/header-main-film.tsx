import * as React from "react";
import {Fragment, ReactElement, FunctionComponent} from "react";
import Header from "../header/header";
import {connect} from "react-redux";
import {getFilm} from "../../store/data/selectors";
import {MAIN_FILM_ID} from "../../constants";
import FilmButtonDiv from "../film-button-div/film-button-div";
import {Film} from "../../type/data";
import {StateApp} from "../../type/reducer";

interface Props {
  film: Film;
}

const HeaderMainFilm: FunctionComponent<Props> = ({
  film
}: Props): ReactElement => {
  if (film) {
    return (
      <Fragment>
        <section className="movie-card">
          <div className="movie-card__bg">
            <img src={film.srcBgImage} alt={film.title} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <Header />

          <div className="movie-card__wrap">
            <div className="movie-card__info">
              <div className="movie-card__poster">
                <img
                  src={film.srcPosterImage}
                  alt={`${film.title} poster`}
                  width="218"
                  height="327"
                />
              </div>

              <div className="movie-card__desc">
                <h2 className="movie-card__title">{film.title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{film.genre}</span>
                  <span className="movie-card__year">{film.released}</span>
                </p>

                <FilmButtonDiv id={film.id} isFavorite={film.isFavorite} />
              </div>
            </div>
          </div>
        </section>
      </Fragment>
    );
  }
  return <Header />;
};
const mapStateToProps = (state: StateApp, ownProps: Props): Props =>
  Object.assign({}, ownProps, {
    film: getFilm(state, MAIN_FILM_ID)
  });

export {HeaderMainFilm};

export default connect(mapStateToProps)(HeaderMainFilm);
