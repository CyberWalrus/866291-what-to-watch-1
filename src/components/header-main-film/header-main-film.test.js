import React from "react";
import renderer from "react-test-renderer";
import {createStore} from "redux";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import {HeaderMainFilm} from "./header-main-film.jsx";
import {FILM} from "../../mock/mock-test.js";
import reducer from "../../store";

it(`HeaderMainFilm correctly renders after relaunch`, () => {
  const store = createStore(reducer);
  const handleClick = jest.fn();
  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <HeaderMainFilm onVideoScreenOpen={handleClick} film={FILM} />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
