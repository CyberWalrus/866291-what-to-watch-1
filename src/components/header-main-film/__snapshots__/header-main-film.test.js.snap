// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderMainFilm correctly renders after relaunch 1`] = `
<section
  className="movie-card"
>
  <div
    className="movie-card__bg"
  >
    <img
      alt="Shutter Island"
      src="https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg"
    />
  </div>
  <h1
    className="visually-hidden"
  >
    WTW
  </h1>
  <header
    className="page-header movie-card__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="user-block"
    >
      <a
        className="user-block__link"
        href="/login"
        onClick={[Function]}
      >
        Sign In
      </a>
    </div>
  </header>
  <div
    className="movie-card__wrap"
  >
    <div
      className="movie-card__info"
    >
      <div
        className="movie-card__poster"
      >
        <img
          alt="Shutter Island poster"
          height="327"
          src="https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg"
          width="218"
        />
      </div>
      <div
        className="movie-card__desc"
      >
        <h2
          className="movie-card__title"
        >
          Shutter Island
        </h2>
        <p
          className="movie-card__meta"
        >
          <span
            className="movie-card__genre"
          >
            Thriller
          </span>
          <span
            className="movie-card__year"
          >
            2010
          </span>
        </p>
        <div
          className="movie-card__buttons"
        >
          <button
            className="btn btn--play movie-card__button"
            onClick={[MockFunction]}
            type="button"
          >
            <svg
              height="19"
              viewBox="0 0 19 19"
              width="19"
            >
              <use
                xlinkHref="#play-s"
              />
            </svg>
            <span>
              Play
            </span>
          </button>
          <button
            className="btn btn--list movie-card__button"
            type="button"
          >
            <svg
              height="20"
              viewBox="0 0 19 20"
              width="19"
            >
              <use
                xlinkHref="#remove"
              />
            </svg>
            <span>
              My list
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;
